python/cugraph/graph/c_graph.pyx

```diff
@@ -280,56 +269,52 @@ class Graph:
         # still in use inside this class. If copy is set to True, deep-copy the
         # objects.
         if copy is False:
-            self.adj_list_offset_col = offset_col;
-            self.adj_list_index_col = index_col;
-            self.adj_list_value_col = value_col;
+            self.adj_list_offset_col = offset_col
+            self.adj_list_index_col = index_col
+            self.adj_list_value_col = value_col
         else:
-            self.adj_list_offset_col = offset_col.copy();
-            self.adj_list_index_col = index_col.copy();
-            self_adj_list_value_col = value_col.copy();
+            self.adj_list_offset_col = offset_col.copy()
+            self.adj_list_index_col = index_col.copy()
+            self_adj_list_value_col = value_col.copy()
 
         cdef uintptr_t graph = self.graph_ptr
-        cdef uintptr_t offsets = create_column(self.adj_list_offset_col)
-        cdef uintptr_t indices = create_column(self.adj_list_index_col)
-        cdef uintptr_t value
+        cdef gdf_column c_offset_col = get_gdf_column_view(self.adj_list_offset_col)
+        cdef gdf_column c_index_col = get_gdf_column_view(self.adj_list_index_col)
+        cdef gdf_column c_value_col
+        cdef gdf_column * c_value_col_ptr
         if value_col is None:
-            value = 0
+            c_value_col_ptr = NULL
         else:
-            value = create_column(self.adj_list_value_col)
+            c_value_col = get_gdf_column_view(self.adj_list_value_col)
+            c_value_col_ptr = &c_value_col
 
-        try:
-            err = gdf_adj_list_view(< gdf_graph *> graph,
-                                    < gdf_column *> offsets,
-                                    < gdf_column *> indices,
-                                    < gdf_column *> value)
-            cudf.bindings.cudf_cpp.check_gdf_error(err)
-        finally:
-            delete_column(offsets)
-            delete_column(indices)
-            if value is not 0:
-                delete_column(value)
+        err = gdf_adj_list_view(<gdf_graph*> graph,
+                                &c_offset_col,
+                                &c_index_col,
+                                c_value_col_ptr)
+        cudf.bindings.cudf_cpp.check_gdf_error(err)
 
     def view_adj_list(self):
         """
         Display the adjacency list. Compute it if needed.
         """
         cdef uintptr_t graph = self.graph_ptr
-        cdef gdf_graph * g = < gdf_graph *> graph
+        cdef gdf_graph * g = <gdf_graph*> graph
         err = gdf_add_adj_list(g)
         cudf.bindings.cudf_cpp.check_gdf_error(err)
 
-        col_size_off = g.adjList.offsets.size
-        col_size_ind = g.adjList.indices.size
+        offset_col_size = self.num_vertices() + 1
```

Comment:
I think it's better to just take the size of the offsets column directly rather than calling self.num_vertices() it's less complicated.

Reply:
I may not agree with this. We'd better hide C++ data structure (gdf_graph) internals from the functions using gdf_graph. If we use num_verticies(), it does not matter whether we change gdf_graph internals or not as long as we have num_verticies(). If some change is necessary, we just need to update num_verticies(). We are about to update graph data structures to match NetworkX, and in this case, we need to search the entire code base and replace every g.adjList.offsets.size -1 (and g.transposedAdjList.offset.size if g.adjList is not available and g.transposedAdjList is available) to match the new data structure. This will become more and more problematic as we add more and more graph functions; there basically are more places we need to track.

See Hiding data (and code) in https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

Hiding data structure internals is a key principle in object oriented programming.

Reply:
And cudf folks are working on replacing gdf_column with cudf::column to provide better abstraction (https://github.com/rapidsai/cudf/issues/1443).

We need to do something similar for gdf_graph; and I think eventually most (if not all) gdf_graph member variables should become private and should not be directly accessed outside gdf_graph.

Reply:
I still disagree. While you are right that it is good to hide the internals of the graph object, the issue is that we aren't looking for the number of vertices in the graph, but for the size of the offsets array. Using `self.num_vertices() + 1` to get that is still relying on information on the graph object's internals, just doing it in a more roundabout way.

Reply:
Thank you for your reply!!!

See "The array IA is of length m + 1." from https://en.wikipedia.org/wiki/Sparse_matrix (here, m is the number of vertices).

self.num_vertices() + 1 (or m + 1 if I use notation from the above link) is from the general knowledge about the CSR format, not from the graph structure internals. Hope this helps to address your concern.

Reply:
While I don't think this should block the PR, I'm mostly concerned about consistency here. It is now using a mix of direct structure access and accessors. If `self.num_vertices()+1` is used to set the size of the offset array, it is best to use a `self.num_edges()` for indices arrays too or keep the previous, consistent, solution imo.

Reply:
Yes, I have the same concern. The thing is that we do not have the num_edges() function, and we are planning to do some major restructuring to match the NetworkX API (for the 0.8 release I guess), so I was a bit hesitant to implement the num_edges() function that will be replaced in one month. I decided to defer this work for the 0.8 release but if this inconsistency really bothers (it somewhat bothers me but I can endure if this will last for only short time), I can do the work for this pull request.

Reply:
Actually, I think I'd better do this now.

I will rename num_vertices() to number_of_nodes() and num_edges() to number_of_edges() to follow NetworkX (https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.DiGraph.number_of_nodes.html and https://networkx.github.io/documentation/stable/reference/classes/generated/networkx.Graph.number_of_edges.html).

This will not be too much work and not gonna be entirely replaced.

Reply:
I can also live with that as long as these changes are tracked in 0.8 release ;)